### ISSUE TITLE
False positive when using sorted(..., key=...): 'cannot be assigned to parameter "key" of type "(_T@sorted) -> SupportsRichComparison"'

### ISSUE THREAD

**Issue by etfdev** [microsoft/pyright#3752]
I expect the following working code to produce no warnings or errors in `pyright`:

```python
import functools
from typing import Callable, Tuple, TypeVar

X = TypeVar("X")
Y = TypeVar("Y")
R = TypeVar("R")

def uncurry2(f: Callable[[X, Y], R]) -> Callable[[Tuple[X, Y]], R]:
    @functools.wraps(f)
    def f_inner(args: Tuple[X, Y]):
        x, y = args
        return f(x, y)

    return f_inner

if __name__ == "__main__":
    tuples = [(1, 2), (5, 6), (3, 4)]
    result1 = sorted(tuples, key=lambda x: x[1])
    def cmp(x: int, y: int) -> int:
        return y
    result2 = sorted(tuples, key=uncurry2(cmp))
    result3 = sorted(tuples, key=uncurry2(lambda x, y: y))
    assert result1 == result2 == result3
    print("success")
```

Yet I get

```
/home/user/poc.py
  /home/user/poc.py:22:34 - error: Argument of type "(Tuple[Literal[3], Literal[4]]) -> SupportsRichComparison" cannot be assigned to parameter "key" of type "(_T@sorted) -> SupportsRichComparison" in function "sorted"
  __Type "(Tuple[Literal[3], Literal[4]]) -> SupportsRichComparison" cannot be assigned to type "(_T@sorted) -> SupportsRichComparison"
  ____Parameter 1: type "_T@sorted" cannot be assigned to type "Tuple[Literal[3], Literal[4]]"
  ______Type "tuple[Literal[1], Literal[2]] | tuple[Literal[5], Literal[6]] | tuple[Literal[3], Literal[4]]" cannot be assigned to type "Tuple[Literal[3], Literal[4]]"
  ________Tuple entry 1 is incorrect type
  __________"Literal[1]" cannot be assigned to type "Literal[3]"
  ________Tuple entry 1 is incorrect type
  __________"Literal[5]" cannot be assigned to type "Literal[3]" (reportGeneralTypeIssues)
  /home/user/poc.py:22:5 - warning: Type of "result2" is unknown (reportUnknownVariableType)
  /home/user/poc.py:23:34 - error: Argument of type "(Tuple[Literal[3], Literal[4]]) -> SupportsRichComparison" cannot be assigned to parameter "key" of type "(_T@sorted) -> SupportsRichComparison" in function "sorted"
  __Type "(Tuple[Literal[3], Literal[4]]) -> SupportsRichComparison" cannot be assigned to type "(_T@sorted) -> SupportsRichComparison"
  ____Parameter 1: type "_T@sorted" cannot be assigned to type "Tuple[Literal[3], Literal[4]]"
  ______Type "tuple[Literal[1], Literal[2]] | tuple[Literal[5], Literal[6]] | tuple[Literal[3], Literal[4]]" cannot be assigned to type "Tuple[Literal[3], Literal[4]]"
  ________Tuple entry 1 is incorrect type
  __________"Literal[1]" cannot be assigned to type "Literal[3]"
  ________Tuple entry 1 is incorrect type
  __________"Literal[5]" cannot be assigned to type "Literal[3]" (reportGeneralTypeIssues)
  /home/user/poc.py:23:5 - warning: Type of "result3" is unknown (reportUnknownVariableType)
2 errors, 2 warnings, 0 informations
Completed in 0.813sec
```

In the non-simplified version of this (the actual production code), the error is a bit different, even though everything is fully typed:

```
Cannot access member "timestamp" for type "object*"
  Member "timestamp" is unknownPylancereportGeneralTypeIssues
```

**VS Code extension or command-line**

I use both.

- VS Code: `Pylance v2022.7.40`
- CLI: `pyright 1.1.227`
- Python: 3.6 and 3.8

**Comment by erictraut** [comment]
This was fixed a couple of weeks ago in pyright, but the fix hasn't yet made it into the production build of pylance. If you switch to the prerelease (insiders) version, you'll see that it's fixed.

**Comment by etfdev** [comment]
Thank you!
